src/sage/rings/function_field/drinfeld_modules/finite_drinfeld_module.py

```diff
@@ -303,31 +512,25 @@ def frobenius_norm(self):
             The Frobenius norm is computed using the formula, by
             Gekeler, given in [MS2019]_, Section 4, Proposition 3.
         """
-        self._check_rank_two()
-        L = self._base.over(self._Fq)
-        # Notations from Schost-Musleh:
-        if self._frobenius_norm is None:
-            n = L.degree_over(self._Fq)
-            d = self.characteristic().degree()
-            m = n // d
-            delta = self._gen[2]
-            norm = L(delta).norm()
-            char = self.characteristic()
-            self._frobenius_norm = ((-1)**n) * (char**m) / norm
+        if self._frobenius_norm is not None:
+            return self._frobenius_norm
+        K = self.base_over_constants_field()
+        n = K.degree(self._Fq)
+        char = self.characteristic()
+        norm = K(self.coefficients()[-1]).norm()
```

Comment:
OK for proceeding like this but I would say that it's worth computing directly the determinant of the matrix acting on the crystalline cohomology, instead of computing this characteristic polynomial and then picking its constant coefficient.
Maybe having a method `_crystalline_matrix(self)` would be a good idea.

Reply:
To elaborate on this remark, maybe we can have something like this:
```sage
    def frobenius_charpoly(self, var='X', algorithm='crystalline'):
        if hasattr(self, f'_frobenius_charpoly_{algorithm}'):
            method = getattr(self, f'_frobenius_charpoly_{algorithm}')
            return method(self, var)
        elif hasattr(self, f'_frobenius_matrix_{algorithm}'):
            method = getattr(self, f'_frobenius_matrix_{algorithm}')
            C = method(self).charpoly()
            return C   # after casting to Fq[X]
        else:
            raise ValueError(f"algorithm '{algorithm}' is not available")

    def frobenius_norm(self, algorithm='crystalline'):
        if hasattr(self, f'_frobenius_matrix_{algorithm}'):
            method = getattr(self, f'_frobenius_matrix_{algorithm}')
            return method(self).det()
        elif hasattr(self, f'_frobenius_charpoly_{algorithm}'):
            method = getattr(self, f'_frobenius_charpoly_{algorithm}')
            C = method(self, 'X')
            return C[0]    # after casting to Fq
        else:
            raise ValueError(f"algorithm '{algorithm}' is not available")

    def frobenius_trace(self, algorithm='crystalline'):
        if hasattr(self, f'_frobenius_matrix_{algorithm}'):
            method = getattr(self, f'_frobenius_matrix_{algorithm}')
            return method(self).trace()
        elif hasattr(self, f'_frobenius_charpoly_{algorithm}'):
            method = getattr(self, f'_frobenius_charpoly_{algorithm}')
            C = method(self, 'X')
            return C[C.degree() - 1]    # after casting to Fq
        else:
            raise ValueError(f"algorithm '{algorithm}' is not available")
```
Then instead of implementing `_frobenius_charpoly_crystalline`, you implement `_frobenius_matrix_crystalline` (which is, of course, part of the code you are already written).

Reply:
By the way, the norm of the Frobenius is always the characteristic to the correct power (namely `n` divided by the degree of the characteristic), maybe up to a sign, isn't it?
If it's true, we can completely shortcut the computation of the characteristic polynomial (or the matrix acting on the cohomology).

Reply:
> By the way, the norm of the Frobenius is always the characteristic to the correct power (namely `n` divided by the degree of the characteristic), maybe up to a sign, isn't it?
> If it's true, we can completely shortcut the computation of the characteristic polynomial (or the matrix acting on the cohomology).

Up to an invertible in $\mathbb{F}_q$, yeah. I believe the exact formula was given by Joseph and Éric in their 2019 paper.

Reply:
So, should we implement it instead?

Reply:
I think it would be a good idea!